CONTRIBUTING.md

```diff
@@ -0,0 +1,76 @@
+this operator.
+
+- Generally, before developing enhancements to this charm, you should consider [opening an issue
+  ](https://github.com/canonical/training-operator/issues) explaining your use case.
+- If you would like to chat with us about your use-cases or proposed implementation, you can reach
+  us at [Canonical Mattermost public channel](https://chat.charmhub.io/charmhub/channels/charm-dev)
+  or [Discourse](https://discourse.charmhub.io/).
+- Familiarising yourself with the [Charmed Operator Framework](https://juju.is/docs/sdk) library
+  will help you a lot when working on new features or bug fixes.
+- All enhancements require review before being merged. Code review typically examines
+  - code quality
+  - test coverage
+  - user experience for Juju administrators of this charm.
+- Please help us out in ensuring easy to review branches by rebasing your pull request branch onto
+  the `main` branch. This also avoids merge commits and creates a linear Git commit history.
+
+## Developing
+
+You can use the environments created by `tox` for development:
+
+```shell
+tox --notest -e unit
+source .tox/unit/bin/activate
+```
+
+### Testing
+
+```shell
+tox -e lint          # code style
+tox -e unit          # unit tests
+tox -e integration   # integration tests
+tox                  # runs 'lint' and 'unit' environments
+```
+
+## Build charm
+
+Build the charm in this git repository using:
+
+```shell
+charmcraft pack
+```
+
+### Deploy
+
+```bash
+# Create a model
+juju add-model dev
+# Enable DEBUG logging
+juju model-config logging-config="<root>=INFO;unit=DEBUG"
+# Deploy the charm
+juju deploy ./training-operator_ubuntu-20.04-amd64.charm \
+    --resource training-operator-image=kubeflow/training-operator
+```
+
+## Charmed Training Operator source and resources upgrade
+
+To upgrade the source and resources Charmed Training Operator, you must:
+
+1) Bump the `training-operator-image` in `metadata.yaml`
+2) Update all manifests in `src/` to match those in the version you want to upgrade to
+3) Ensure integration and unit tests are passing; fix/adapt them otherwise
+
+The following table collects a series of links and information that can help in the upgrade process.
+ 
+| Image repository 	| Image version 	| Manifest location 	| Next version 	|
+|---	|---	|---	|---	|
+| [kubeflow/training-operator](https://hub.docker.com/r/kubeflow/training-operator/) 	| `v1-<commit sha>`<br>The commit sha comes from the HEAD of the release tag 	| [kubeflow/training-operator/manifests](https://github.com/kubeflow/training-operator/tree/v1.5.0-rc.0/manifests) 	| Check for latest version released or pre-released in <br>the repo [releases](https://github.com/kubeflow/training-operator/releases), it should match the release plan usually stated in issues like <br>[Training Operator WG and Kubeflow 1.6](https://github.com/kubeflow/manifests/issues/2196) release and [KF Release 1.6 Tracking](https://github.com/kubeflow/manifests/issues/2194) 	|
```

Comment:
### Re `image repository` column:

I think now I might understand some of our difference both for this and `Image version`... When you're looking to update Training Operator, where do you look for info?  I look at [kubeflow/manifests](https://github.com/kubeflow/manifests) (if we're updating to sync with a new kubeflow), or [training operators/manifest](https://github.com/kubeflow/training-operator/tree/master/manifests) (if we're updating an application on its own, which usually we don't do, but maybe it'll happen sometimes).  And I get the feeling you're going to the container registry to get the image tag, is that correct?  Maybe that's why we're feeling differently about the info

Reply:
I generally look at something like `values.yaml` in manifests to see what tag the manifests are pointing. Having a link to the registry still might be interesting. But I would rename this from `image repository` to `image registry`

Reply:
Yeah I look at manifests, not the image repo.  Manifests have the image version that should be used, but also context on how to use it.  The image version alone doesn't say much

And then if I'm picking between manifests, agreed `repo/manifests` is good for manifests deploying that application, but since we're mostly deploying a suite of interconnected applications I think the `kubeflow/manifest` repo is probably more relevant to us most of the time (otherwise maybe we'll deploy a training-operator that breaks some other thing in kubeflow)